### ISSUE TITLE
Confusion matrix

### ISSUE THREAD

**Issue by tagny** [lengocluyen/relation_extraction_textmine25#5]
- ~~sklearn [multilabel_confusion_matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.multilabel_confusion_matrix.html)~~ _broken link_ _multiple binary confusion matrices_
- 2020 [MULTI-LABEL CLASSIFIER PERFORMANCE EVALUATION WITH CONFUSION MATRIX](https://csitcp.com/paper/10/108csit01.pdf)

````python
import pandas as pd
import numpy as np

Y = np.array([[1, 1, 0, 0], [0, 1, 1, 0], [0, 0, 0, 1], [1, 1, 1, 1], [0, 1, 1, 0], [0, 1, 1, 0], [0, 1, 0, 1]])  # GT
Z = np.array([[1, 1, 0, 0], [1, 1, 1, 0], [1, 0, 0, 1], [0, 1, 1, 1], [0, 1, 0, 0], [1, 1, 0, 0], [1, 0, 1, 0]]) # predictions

q = Y.shape[1]  # number of labels
N = Y.shape[0]  # number of examples

M = np.zeros(shape=(q, q))
for i in range(N):
    if all(Y[i] == Z[i]):
        print(i, "case i")
        M += np.diag(Y[i])
    else:
        yi = set(np.argwhere(Y[i] > 0).flatten().tolist())
        zi = set(np.argwhere(Z[i] > 0).flatten().tolist())
        yi_diff_zi_arr = np.zeros(q)
        yi_diff_zi_arr[list(yi.difference(zi))] = 1
        zi_diff_yi_arr = np.zeros(q)
        zi_diff_yi_arr[list(zi.difference(yi))] = 1
        yi_inter_zi_arr = np.zeros(q)
        yi_inter_zi_arr[list(yi.intersection(zi))] = 1
        if len(yi.difference(zi)) == 0 and len(zi.difference(yi)) > 0:
            print(i, "case ii")
            M += (np.outer(yi_inter_zi_arr, zi_diff_yi_arr) + len(yi) * np.diag(Y[i])) / len(zi)
        elif len(yi.difference(zi)) > 0 and len(zi.difference(yi)) == 0:
            print(i, "case iii")
            M += np.outer(yi_diff_zi_arr, Z[i])/len(zi) + np.diag(Z[i])
        else:
            print(i, "case iv")
            M += np.outer(yi_diff_zi_arr, zi_diff_yi_arr)/len(zi.difference(yi)) + np.diag(yi_inter_zi_arr)
````

- 2022 [MLCM: Multi-Label Confusion Matrix](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9711932)
  - ~~[mlcm library](https://pypi.org/project/mlcm/)~~ _erroneous_
- 2023 [Comments on ‘‘MLCM: Multi-Label Confusion Matrix’’](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=10103535)
